src/main/java/htsjdk/samtools/util/EdgeReadIterator.java

```diff
@@ -88,73 +139,86 @@ public EdgeReadIterator(final SamReader samReader, final IntervalList intervalLi
+        // have to find intersections and can later emit the read right away.
+        final boolean needToConsiderIntervals = intervals != null && !intervalCompletelyContainsRead(rec);
+
         // interpret the CIGAR string and add the base info
         for (final AlignmentBlock alignmentBlock : rec.getAlignmentBlocks()) {
             // 0-based offset into the read of the current base
-            final int readOffset = alignmentBlock.getReadStart() - 1;
+            final int offsetStartOfAlignmentBlockInRead = alignmentBlock.getReadStart() - 1;
             // 1-based reference position that the current base aligns to
-            final int refPos = alignmentBlock.getReferenceStart();
+            final int referencePositionStartOfAlignmentBlock = alignmentBlock.getReferenceStart();
 
+            // Here we add the first entry to the accumulator, which is the start of this AlignmentBlock.
             if (accumulator.isEmpty()) {
                 accumulator.add(createLocusInfo(getReferenceSequence(rec.getReferenceIndex()), rec.getAlignmentStart()));
             }
 
-            // The accumulator should always have LocusInfos that correspond to one consecutive segment of loci from one reference
-            // sequence. So
+            // The accumulator should always have LocusInfos that correspond to one consecutive segment of loci from
+            // one reference sequence. So
             // accumulator.get(0).getPosition() + accumulator.size() == accumulator.get(accumulator.size()-1).getPosition()+1
             final int accumulatorNextPosition = accumulator.get(0).getPosition() + accumulator.size();
 
             if (accumulatorNextPosition != accumulator.get(accumulator.size() - 1).getPosition() + 1) {
                 throw new IllegalStateException("The accumulator has gotten into a funk. Cannot continue");
             }
 
-            // Ensure there are AbstractLocusInfos up to and including this position
-            for (int locusPos = accumulatorNextPosition; locusPos <= refPos + alignmentBlock.getLength(); ++locusPos) {
+            // Ensure there are consecutive AbstractLocusInfos up to and including the end of the AlignmentBlock
+            for (int locusPos = accumulatorNextPosition; locusPos <= referencePositionStartOfAlignmentBlock + alignmentBlock.getLength(); ++locusPos) {
                 accumulator.add(createLocusInfo(getReferenceSequence(rec.getReferenceIndex()), locusPos));
             }
 
-            /* Let's assume an alignment block starts in some locus. 
-             * We put two records to the accumulator. The first one has the "begin" type which corresponds to the locus 
-             * where the block starts. The second one has the "end" type which corresponds to the other locus where the block ends. 
-            */
+            // Let's assume an alignment block starts in some locus.
```

Comment:
This is all pretty confusing.  It could really use a nice ASCII art if you're feeling creative.

Reply:
```\o/```